### DISCUSSION TITLE
Multiple Viewports without crashing?

### DISCUSSION THREAD

**Discussion by 11Spades** [AllenDang/giu#906] [Q&A] [answered]
As I understand it, using multiple viewports is kind-of shaky in imgui, especially in linux, but I'd really like to do it anyway. Unfortunately, it isn't very clear to me how I can actually achieve this. I've been trying to glue something together via `imgui`'s examples, and then praying they translate well, but this causes crashes right now.

Currently, my approach is as follows:
Before the main loop, invoke:
```go
imgui.IO.SetConfigFlags(*imgui.CurrentIO(), imgui.ConfigFlagsViewportsEnable)
```

Then, at the end of the main loop, do this:
```go
imgui.UpdatePlatformWindows()
imgui.RenderPlatformWindowsDefault()
```

I believe that the crucial parts missing here, at least according to the GLFW-backend examples of how to do this in `imgui`'s examples, are as such:
`GLFWwindow* backup_current_context = glfwGetCurrentContext();` -- I can't seem to figure out how to do this in GIU or it's imgui component, nor its GLFW backend component.
`glfwMakeContextCurrent(backup_current_context);` -- I also can't find the corresponding call for this.

I'm also unsure whether the backend is platform-dependent; while I am on Linux, the tools I am building are for people who run Windows and MacOS, so perhaps I will need to call different functions depending on the platform?

Guidance would be appreciated!

**Comment by gucio321** [comment, marked as answer]
Generally you shouldn't need all of that as `ViewportsEnable` is set by default in glfw backend

**Reply by 11Spades** [reply]
Ah. I feel a bit stupid :P

Thank you for the help.

Somewhat offtopic, but is there a way to ensure that any new viewports are transparent if the master window is transparent?

**Reply by gucio321** [reply]
Idk to be honest, could you show an example when where it isn't

**Reply by 11Spades** [reply]
As of right now, I'm achieving transparency by doing this outside of the main loop:
```go
	mainPane := g.NewMasterWindow("Map", 800, 600, g.MasterWindowFlagsFloating|g.MasterWindowFlagsFrameless|g.MasterWindowFlagsTransparent) // No resize is recommended, but doesn't actually seem to be necessary, nor does it solve the problem.
	mainPane.SetBgColor(color.Transparent)
```

and then, in the main loop, I'm doing this:
```go
	// Set up transparency
	imgui.PushStyleVarFloat(imgui.StyleVarWindowBorderSize, 0)
	g.PushColorWindowBg(color.RGBA{0, 0, 0, 150})
	g.PushColorFrameBg(color.RGBA{0, 0, 0, 0})
...
	// This is transparent only when docked in the main viewport
	for i, windowData := range subWindows {
		window := g.Window(windowData.From + " -> " + windowData.To)
		window.Size(vX/5, vY/3)
		window.Layout(...)
	}

	// This is transparent unconditionally, because it is always in the main window
	mainWindow := g.Window("Lodestar")
	mainWindow.Flags(g.WindowFlagsNoDecoration | g.WindowFlagsNoMove | g.WindowFlagsNoBringToFrontOnFocus)
	mainWindow.Size(vX, vY)
	mainWindow.Layout(roomGraph)
...
	// Pop our styles so we don't crash
	g.PopStyleColor()
	g.PopStyleColor()
	g.PopStyle()

```

In this case, my suspicion is that the g.MasterWindowFlagsTransparent flag exclusively applies to the main viewport, and there doesn't seem to be a way to contact GLFW regarding child windows to set the transparent flag on them.
